hypothesis/pip-sync-faster
# Re-add non-boilerplate code and add tests

This re-adds the non-boilerplate files that https://github.com/hypothesis/pip-sync-faster/pull/5 (which this PR is based on) removed in order to facilitate code review. Also adds tests.

Once again a review of the meat is required (`README.md`, `src/*` and `tests/*`) and the cookiecutter stuff (`.cookiecutter/*`) can be reviewed when we review the cookiecutter itself.

# How to test manually

The pip-sync-faster concept is pretty simple: it calls `pip-sync` and stashes hashes of the requirements files in a `pip_sync_faster.json` file within the venv. If you run it again with the same requirements files and you haven't changed any of the requirements files then it won't call `pip-sync` again. If any of the requirements files has changed or if you call it with a different set of requirements files (even a subset of the ones you called it with last time) then it will call `pip-sync` again.

The CLI of `pip-sync-faster` is the same as that of `pip-sync` and it passes all command line options and arguments blindly to `pip-sync`. It also echoes `pip-sync`'s exit status.

To test it:

* Create and activate a venv

* Create some test requirements files

* Install `pip-sync-faster` in the venv:

  ```terminal
  pip install -e /path/to/pip-sync-faster
  ```

* It should be pretty obvious to see that it's running pip-sync:

  ```terminal
  pip-sync-faster requirements.txt
  ```

* There is one testing gotcha: `pip-sync-faster` must itself be in the requirements file, otherwise it'll uninstall itself (it'll call `pip-sync requirements.txt` which'll uninstall anything that's not in `requirements.txt`, including `pip-sync-faster`).

  In testing this is a double-gotcha because even if `requirements.txt` contains `pip-sync-faster` it will replace the local development version with a copy from PyPI.

  So in testing a local development version you have to reinstall it every time:

  ```terminal
  pip install -e /path/to/pip-sync-faster
  ```

* If you run `pip-sync-faster` again with the same requirements file it won't call `pip-sync`:

  ```terminal
  pip-sync-faster requirements.txt
  ```

Other things to test:

* If you modify the requirements file and then re-call `pip-sync-faster` it will call `pip-sync`

* If you call it with a different requirements file it will call `pip-sync`. If you then call it with the first requirements file again it'll call `pip-sync` again

* Calling it with multiple requirements files. If you call it again without changing any of the requirements files it won't call `pip-sync`. If you change any of them it will

* If you call it with a different set of requirements files (even a subset of what you previously called it with) it will call `pip-sync`
